### PR TITLE
WT-6598 Add API to set number of hash buckets.

### PR DESCRIPTION
**Pull request by sueloverso**

### CODE REVIEW

**Comment by sueloverso** [comment]
@keitharnoldsmith and @ddanderson please review this. @agorrod also FYI. I added official stats for the various components I was timing during my testing, during checkpoint's gathering of handles (while holding the schema lock) thinking it could help give us insight into what customers loads look like when they have thousands of handles.

I've gone back and forth a bit on naming too. So if you don't like something let me know. 

One thing that slightly bothers me is that it would be easy for someone in the future to use the wrong hash size field for the wrong bucket array. I.e. using `conn->hash_size` for a dhandle list and not finding their handle and inserting it into a 2nd bucket, or the other way around. I could not think of a good way to code fail-safes into there.

**Comment by ddanderson** [comment]
@sueloverso, as a followup to my previous comment.  If there is any performance difference another option to consider would be to compute modulo via  `bucket = hash_value & (S2C(session)->hash_size - 1)`.  That certainly requires `hash_size` to be a power of 2, which would need to be enforced by the API.

File: dist/stat_data.py [hunk 1]
```diff
@@ -622,6 +624,12 @@ def __init__(self, name, desc, flags=''):
     TxnStat('txn_checkpoint', 'transaction checkpoints'),
     TxnStat('txn_checkpoint_fsync_post', 'transaction fsync calls for checkpoint after allocating the transaction ID'),
     TxnStat('txn_checkpoint_fsync_post_duration', 'transaction fsync duration for checkpoint after allocating the transaction ID (usecs)', 'no_clear,no_scale'),
+    TxnStat('txn_checkpoint_handle_applied', 'transaction checkpoint most recent handles applied'),
+    TxnStat('txn_checkpoint_handle_skipped', 'transaction checkpoint most recent handles skipped'),
+    TxnStat('txn_checkpoint_handle_walked', 'transaction checkpoint most recent handles walked'),
+    TxnStat('txn_checkpoint_handle_duration', 'transaction checkpoint most recent duration for gathering handles (usecs)', 'no_clear,no_scale'),
```
[the remark this answers is not in the document]
Comment (sueloverso) [reply dist/stat_data.py:630]: Fixed.

File: dist/api_data.py [hunk 2]
```diff
@@ -1016,6 +1016,17 @@ def __ge__(self, other):
...
         type='boolean'),
+    Config('hash', '', r'''
+        manage resources around hash bucket arrays. Note that setting large values can
+        significantly increase memory usage inside WiredTiger''',
+        type='category', subconfig=[
+        Config('buckets', 512, r'''
+            configure the general number of hash buckets for most system hash arrays''',
+            min='64', max='65536'),
+        Config('dhandle_buckets', 512, r'''
+            configure the number of hash buckets for arrays relating to data handles''',
```
[the remark this answers is not in the document]
Comment (sueloverso) [reply dist/api_data.py:1027]: Fixed.

**Comment by sueloverso** [comment]
@ddanderson I don't think it is a burden to require the value be a power of 2. What cursor caching test should I pay attention to? I don't think cursor caching is part of any of the standard wtperf performance plots.

File: dist/api_data.py [hunk 3]
```diff
@@ -1016,6 +1016,17 @@ def __ge__(self, other):
         modified file is always flushed to storage when closing file handles to
         acquire exclusive access to the table''',
         type='boolean'),
+    Config('hash', '', r'''
+        manage resources around hash bucket arrays. Note that setting large values can
+        significantly increase memory usage inside WiredTiger''',
+        type='category', subconfig=[
+        Config('buckets', 512, r'''
+            configure the general number of hash buckets for most system hash arrays''',
```
[the remark this answers is not in the document]
Comment (sueloverso) [reply dist/api_data.py]: Fixed

**Comment by sueloverso** [comment]
For starters, I pointed the Jenkins perf jobs at my branch and kicked off the perf tests.

**Comment by sueloverso** [comment]
FYI - the PR failure is rollback_to_stable10, a known failure.

**Pull request merged**